### ISSUE TITLE
`tf.range` have accumulate floating point error on CPU

### ISSUE THREAD

**Issue by fuzzyswan** [tensorflow/tensorflow#58133]
<details><summary>Click to expand!</summary> 
 
 ### Issue Type

Bug

### Source

binary

### Tensorflow Version

tf 2.10.0

### Custom Code

Yes

### OS Platform and Distribution

Linux Ubuntu 20.04

### Current Behaviour?

`tf.range` have accumulate greater and greater floating point error on CPU. In the example below, `range(- 10, 10, 0.01)` gives `[... 9.980267 9.990267]` when the correct result should be `[... 9.98 9.99]`, and this deviation is too much in my opinion.

### Standalone code to reproduce the issue

```shell
import tensorflow as tf
with tf.device('cpu'):
    tensor1 = tf.range(- 10, 10, 0.01) # [-10., -9.99 ... 9.980267 9.990267]
    print(tensor1)
with tf.device('gpu'):
    tensor2 = tf.range(- 10, 10, 0.01) # [-10., -9.99 ... 9.98 9.99]
    print(tensor2)
assert np.allclose(tensor1, tensor2) # AssertionError
```

### Relevant log output

```shell
tf.Tensor([-10.        -9.99      -9.98     ...   9.970266   9.980267   9.990267], shape=(2000,), dtype=float32)
tf.Tensor([-10.        -9.99      -9.98     ...   9.969999   9.98       9.99    ], shape=(2000,), dtype=float32)
AssertionError
```
</details>

**Comment by tilakrayal** [comment]
@fuzzyswan,
I was able to execute the given code without any issue/error. Kindly find the gist of it [here](https://colab.research.google.com/gist/tilakrayal/1fe30e38b088b5344b9a039caf5747b1/untitled695.ipynb). Thank you!

**Comment by fuzzyswan** [comment]
Hi @tilakrayal , I run on TensorFlow 2.10 and turn on GPU on colab, and was able to reproduce the issue/error. Please find the [gist](https://colab.research.google.com/gist/fuzzyswan/739cca41a028bd92bde975dce8494a6b/untitled695.ipynb) here, thanks!

**Comment by bhack** [comment]
@fuzzyswan Same as https://github.com/tensorflow/tensorflow/issues/58131#issuecomment-1283098755 probably they are confused that the soft placement is enabled by default. @tilakrayal You are running on a CPU colab (test it your gist with ```tf.config.set_soft_device_placement(False)```)

**Comment by fuzzyswan** [comment]
@bhack The assertion indeed passes with `dtype=tf.float64` and both CPU and GPU produce accurate results.

Although this issue can be avoided in this way (by using a higher precision data type), I just wonder if the implementation for `tf.range` could be improved because now the accumulated error seems very large on CPU, while on GPU it can produce relatively accurate results even with `float32`.

**Comment by bhack** [comment]
It is in the natural c++ float precision limits. What we could do is to align CPU and GPU impl.

See https://github.com/tensorflow/tensorflow/pull/58259

**Comment by bhack** [comment]
You can check yourself in pure c++ replicating your example with the CPU version in master vs the GPU aligned version.

```c++
#include <iostream>

using namespace std;

int main()
{
   // TF master
    float val = -10.f;
    for (int64_t i=0; i<2000; ++i) {
        val+=0.01f;
        cout << "A " << val << endl;
    }
    
    // PR #58259 
     for (int64_t i=0; i<2000; ++i) {
        val=-10.f + static_cast<float>(i) * 0.01;
        cout << "B " << val << endl;
    }
}
```

**Comment by bhack** [comment]
If you are interested more in general check:
https://en.wikipedia.org/wiki/Kahan_summation_algorithm

**Comment by fuzzyswan** [comment]
@bhack thanks for your PR and for providing useful references to numerical analysis. I did not know the summation algorithm before and it is very clever and interesting to me!

I searched in the TensorFlow issues and found a related unresolved problem https://github.com/tensorflow/tensorflow/issues/51419, so I wonder if [Kahan_summation_algorithm](https://en.wikipedia.org/wiki/Kahan_summation_algorithm) is actually used in tensorflow implementation for popupar APIs like `range` and `reduce_sum` to avoid such numerical instability, as I notice that pytorch has adopted cascade-summation algorithm for floats to avoid numerical instability https://github.com/pytorch/pytorch/pull/39516.
